### PR TITLE
Create spans late

### PR DESCRIPTION
**Pull request by Restioson**
This PR changes the instrumentation to create spans only after first send poll. This allows us to instrument the futures from send with `.send(...).instrument(parent)` properly, having the send span come under the `parent` span.

### CODE REVIEW

File: src/envelope.rs [hunk 1]
```diff
@@ -72,53 +56,61 @@ impl HandlerSpan {
...
 
-    fn disabled() -> HandlerSpan {
+    fn none() -> Span {
         #[cfg(feature = "instrumentation")]
-        let span = HandlerSpan(tracing::Span::none());
+        let span = Span(tracing::Span::none());
 
         #[cfg(not(feature = "instrumentation"))]
-        let span = HandlerSpan();
+        let span = Span();
```
Comment (thomaseizinger) [inline src/envelope.rs:64]: It would make the review easier to not have this rename in here, or at least in a commit at the very front of the PR :)

File: src/inbox.rs [hunk 2]
```diff
@@ -89,7 +89,7 @@ impl<A> ChanInner<A> {
         // If len < cap after popping this message, try fulfill at most one waiting sender
         if capacity.map_or(false, |cap| cap == self.ordered_queue.len()) {
             match self.try_fulfill_sender(MessageType::Ordered) {
-                Some(SentMessage::ToOneActor(msg)) => self.ordered_queue.push_back(msg.val),
+                Some(MessageKind::ToOneActor(msg)) => self.ordered_queue.push_back(msg.val),
```
Comment (thomaseizinger) [inline src/inbox.rs:92]: Same with this rename!

File: src/envelope.rs [hunk 3]
```diff
@@ -21,47 +21,31 @@ pub trait MessageEnvelope: Send {
     /// The type of actor that this envelope carries a message for
     type Actor;
 
-    /// Handle the message inside of the box by calling the relevant `AsyncHandler::handle` or
-    /// `Handler::handle` method, returning its result over a return channel if applicable. The
-    /// reason that this returns a future is so that we can propagate any `Handler` responder
-    /// futures upwards and `.await` on them in the manager loop. This also takes `Box<Self>` as the
+    /// Starts the instrumentation of this message request. This will create the request span.
+    fn start_span(&mut self, msg_name: &'static str);
```
Comment (thomaseizinger) [inline src/envelope.rs:25]: Is this really the best way we can implement this?

From reading the PR description "create span on first poll", I did not expect something like this but a change to `poll` of `SendFuture`. I haven't tried implementing this myself so perhaps this is impossible but I am finding the indirection here quite confusing.

Also not really happy with the amount of conditional code that we are growing here.

**Review by thomaseizinger — COMMENTED** [review]
Some initial thoughts.

I think we should support what is outlined in the PR description but I find it quite hard to follow the implementation atm.

I'll play around with this myself and see what I can come up with!

File: src/inbox.rs (hunk 2 above)
Comment (Restioson) [reply src/inbox.rs:92]: This was done since SentMessage became a struct with more stuff inside of it

File: src/envelope.rs (hunk 3 above)
Comment (Restioson) [reply src/envelope.rs:25]: > From reading the PR description "create span on first poll", I did not expect something like this but a change to poll of SendFuture. I haven't tried implementing this myself so perhaps this is impossible but I am finding the indirection here quite confusing.

I thought about this but wasn't entirely sure what the best design would be for this so I prototyped it this way (part of the work was getting the semantics right about when the span should actually be created).

One early concept I had was storing the message in `SendFuture` in a new `New` variant of the inner type. However, this refers to the actor type, so I wasn't sure whether this would work with the `sending_erased` version.

> Also not really happy with the amount of conditional code that we are growing here.

Are you referring to feature-conditional/cfg code or branches?

**Comment by thomaseizinger** [comment]
I have a branch locally that moves some stuff around in the hope that features like this will be easier to build.

I am happy to merge this for now IF we can add a test that ensures my upcoming refactoring is not gonna break anything 😁

**Comment by thomaseizinger** [comment]
> > I am happy to merge this for now IF we can add a test that ensures my upcoming refactoring is not gonna break anything
> 
> Hmm, I wonder how we can write tests for tracing. I don't really have experience with testing for side-effects of the program like logs! I can look into this tomorrow.

I wrote a custom subscriber here at some point to test log output: https://github.com/comit-network/xmr-btc-swap/blob/master/swap/src/tracing_ext.rs

File: src/envelope.rs (hunk 3 above)
Comment (thomaseizinger) [reply src/envelope.rs:25]: It would be nice if we could flatten out the layers in such a way that we can start the instrumentation at a point where we still now the message and actor types but are still outside the `SendFuture`.

> > From reading the PR description "create span on first poll", I did not expect something like this but a change to poll of SendFuture. I haven't tried implementing this myself so perhaps this is impossible but I am finding the indirection here quite confusing.
> 
> I thought about this but wasn't entirely sure what the best design would be for this so I prototyped it this way (part of the work was getting the semantics right about when the span should actually be created).
> 
> One early concept I had was storing the message in `SendFuture` in a new `New` variant of the inner type. However, this refers to the actor type, so I wasn't sure whether this would work with the `sending_erased` version.
> 
> > Also not really happy with the amount of conditional code that we are growing here.
> 
> Are you referring to feature-conditional/cfg code or branches?

Feature-conditional code.

**Comment by Restioson** [comment]
I've added a test now

File: tests/tracing.rs [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    type Return = ();
+
+    async fn handle(&mut self, message: Hello, _ctx: &mut Context<Self>) {
+        tracing::info!("Hi {}", message.0)
+    }
+}
+
+#[tokio::test]
+async fn assert_send_is_child_of_span() {
+    let mock_writer = MockWriter::new(&GLOBAL_BUF);
```
Comment (thomaseizinger) [inline tests/tracing.rs:130]: Do we need to use a global buffer here? Looks trivial to use a function-local buffer which would allow us to write more of these tests that can be run in parallel, plus we can drop the lazy_static dependency.

File: tests/tracing.rs [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+    let (addr, mut ctx) = Context::<Tracer>::new(None);
+    let _ = addr
+        .send(Hello("hi"))
+        .split_receiver()
+        .instrument(tracing::info_span!("Test span"))
+        .await;
+    ctx.yield_once(&mut Tracer).await;
+
+    logs_assert(|lines: &[&str]| lines[0].contains("Test span:xtra_actor_request"));
```
Comment (thomaseizinger) [inline tests/tracing.rs:142]: It would be nice to assert the entire log line here :)

We can disable timestamps to make sure the test is not dependent on time.

File: tests/tracing.rs (hunk 4 above)
Comment (Restioson) [reply tests/tracing.rs:130]: Good point!

File: tests/tracing.rs (hunk 5 above)
Comment (Restioson) [reply tests/tracing.rs:142]: We could do this, but I'm not sure if the format is semver guaranteed?

Comment (thomaseizinger) [reply tests/tracing.rs:142]: I would expect them to make an effort to not break it and even if it does, fixing the test seems like a small effort for the benefit of a more readable assertion (that for example contains the actual log message too).

File: tests/tracing.rs (hunk 4 above)
Comment (thomaseizinger) [reply tests/tracing.rs]: If we wrap the buffer in a new type, we can hide the details of converting the bytes to a lines iterator.

Comment (Restioson) [reply tests/tracing.rs]: Is there much point when this will mostly just be used through `with_logs` anyway?

**Comment by Restioson** [comment]
Uh oh, committed with someone else as author :sweat_smile: let me take this opportunity to squash some stuff up

File: tests/tracing.rs [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+        .split_receiver()
+        .instrument(tracing::info_span!("Test span"))
+        .await;
+    ctx.yield_once(&mut Tracer).await;
+
+    with_logs(&buf, |lines: &[&str]| {
+        assert_eq!(
+            lines,
+            [
+                r#" INFO Test span:xtra_actor_request{actor="tracing::Tracer" tracing::Hello}:xtra_message_handler{actor="tracing::Tracer" tracing::Hello}: tracing: Hi hi"#
```
Comment (thomaseizinger) [inline tests/tracing.rs:142]: I think we can actually see here that using `message` as a span variable name has weird side effects. Should maybe call it `msg_type` or something?

File: tests/tracing.rs [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+    let buf = Arc::new(Mutex::new(vec![]));
+    let mock_writer = MockWriter::new(buf.clone());
+    let subscriber = get_subscriber(mock_writer, "tracing=trace,xtra=trace");
+    let _g = tracing::dispatcher::set_default(&subscriber);
+
+    let (addr, mut ctx) = Context::<Tracer>::new(None);
+    let _ = addr
+        .send(Hello("hi"))
+        .split_receiver()
+        .instrument(tracing::info_span!("Test span"))
```
Comment (thomaseizinger) [inline tests/tracing.rs:134]: Can we use a span name here without spaces? That would make it easier to spot in the assertion.

File: tests/tracing.rs (hunk 6 above)
Comment (thomaseizinger) [inline tests/tracing.rs:142]: I think the last "tracing" here is the test's name right? Can we rename that too? It is a bit confusing to read so many things named tracing :)

Comment (Restioson) [reply tests/tracing.rs:143]: That's a good point, actually. It seems to work okay with Jaeger (and Grafana Tempo), so I never noticed this

File: tests/tracing.rs (hunk 7 above)
Comment (thomaseizinger) [reply tests/tracing.rs:135]: Maybe call it `user_span` or something?

File: tests/tracing.rs (hunk 6 above)
Comment (Restioson) [reply tests/tracing.rs:143]: No, the last `tracing` is the log target, which defaults to the module

Comment (Restioson) [reply tests/tracing.rs]: Speaking of this, I was thinking of removing the tags from everything but `xtra_actor_request`, otherwise they repeat. Should this and the rename go in a separate PR?

Comment (Restioson) [reply tests/tracing.rs]: We could just override it, I think

File: tests/tracing.rs [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+async fn assert_send_is_child_of_span() {
+    let buf = Arc::new(Mutex::new(vec![]));
+    let mock_writer = MockWriter::new(buf.clone());
+    let subscriber = get_subscriber(mock_writer, "tracing=trace,xtra=trace");
+    let _g = tracing::dispatcher::set_default(&subscriber);
+
+    let (addr, mut ctx) = Context::<Tracer>::new(None);
+    let _ = addr
+        .send(Hello("hi"))
+        .split_receiver()
```
Comment (thomaseizinger) [inline tests/tracing.rs:134]: I find the usage of `split_receiver` a bit noisy here. Why not just use `spawn_global` here?

Using `split_receiver` here seems like it matters for the test.

File: tests/tracing.rs [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+#[tokio::test]
+async fn assert_send_is_child_of_span() {
+    let buf = Arc::new(Mutex::new(vec![]));
+    let mock_writer = MockWriter::new(buf.clone());
+    let subscriber = get_subscriber(mock_writer, "tracing=trace,xtra=trace");
+    let _g = tracing::dispatcher::set_default(&subscriber);
+
+    let (addr, mut ctx) = Context::<Tracer>::new(None);
+    let _ = addr
+        .send(Hello("hi"))
```
Comment (thomaseizinger) [inline tests/tracing.rs:133]: Nit:


```suggestion
        .send(Hello("world"))
```

File: tests/tracing.rs (hunk 6 above)
Comment (thomaseizinger) [reply tests/tracing.rs]: Right, I was meant to say it is the test _file_'s name.

File: tests/tracing.rs (hunk 8 above)
Comment (Restioson) [reply tests/tracing.rs:134]: > I find the usage of split_receiver a bit noisy here. Why not just use spawn_global here?

The default subscriber is thread local, I believe, so we need to handle the message on the same thread. It might also change with tokio task - I'm not sure. To make sure everything definitely worked, I wanted to do everything on the same task and thread.

File: tests/tracing.rs (hunk 9 above)
Comment (Restioson) [reply tests/tracing.rs:133]: :laughing:

File: tests/tracing.rs (hunk 6 above)
Comment (thomaseizinger) [reply tests/tracing.rs]: > Speaking of this, I was thinking of removing the tags from everything but `xtra_actor_request`, otherwise they repeat. Should this and the rename go in a separate PR?

I'd prefer a 2nd PR yeah :)

File: tests/tracing.rs (hunk 8 above)
Comment (thomaseizinger) [reply tests/tracing.rs:134]: > > I find the usage of split_receiver a bit noisy here. Why not just use spawn_global here?
> 
> The default subscriber is thread local, I believe, so we need to handle the message on the same thread. It might also change with tokio task - I'm not sure. To make sure everything definitely worked, I wanted to do everything on the same task and thread.

I believe the default tokio test executor is single-threaded and if not, you can customise it via the `tokio::test` attribute.

File: tests/tracing.rs (hunk 6 above)
Comment (Restioson) [reply tests/tracing.rs]: I've renamed it to `instrumentation`

File: tests/tracing.rs (hunk 8 above)
Comment (thomaseizinger) [reply tests/tracing.rs]: Yep, it is single-threaded: https://docs.rs/tokio/latest/tokio/attr.test.html#using-default

Comment (Restioson) [reply tests/tracing.rs]: I've changed it now

**Comment by Restioson** [comment]
Merged with master now
